src/scheduler/worker.rs

```diff
@@ -140,22 +145,131 @@ impl<VM: VMBinding> GCWorker<VM> {
-        self.shared.parked.store(false, Ordering::SeqCst);
         loop {
-            while let Some((bucket, mut work)) = self.local_work_buffer.pop() {
-                debug_assert!(self.scheduler.work_buckets[bucket].is_activated());
-                work.do_work_with_stat(self, mmtk);
-            }
-            let mut work = self.scheduler().poll(self);
-            debug_assert!(!self.shared.is_parked());
+            let mut work = self.poll();
             work.do_work_with_stat(self, mmtk);
         }
     }
 }
+
+/// A worker group to manage all the GC workers (except the coordinator worker).
+pub struct WorkerGroup<VM: VMBinding> {
+    /// Shared worker data
+    pub workers_shared: Vec<Arc<GCWorkerShared<VM>>>,
+    parked_workers: AtomicUsize,
+    unspawned_local_work_queues: Mutex<Vec<deque::Worker<Box<dyn GCWork<VM>>>>>,
+}
+
+impl<VM: VMBinding> WorkerGroup<VM> {
+    /// Create a WorkerGroup
+    pub fn new(num_workers: usize) -> Arc<Self> {
+        let unspawned_local_work_queues = (0..num_workers)
+            .map(|_| deque::Worker::new_fifo())
+            .collect::<Vec<_>>();
+
+        let workers_shared = (0..num_workers)
+            .map(|i| {
+                Arc::new(GCWorkerShared::<VM>::new(Some(
+                    unspawned_local_work_queues[i].stealer(),
+                )))
+            })
+            .collect::<Vec<_>>();
+
+        Arc::new(Self {
+            workers_shared,
+            parked_workers: Default::default(),
+            unspawned_local_work_queues: Mutex::new(unspawned_local_work_queues),
+        })
+    }
+
+    /// Spawn all the worker threads
+    pub fn spawn(
+        &self,
+        mmtk: &'static MMTK<VM>,
+        sender: Sender<CoordinatorMessage<VM>>,
+        tls: VMThread,
+    ) {
+        let mut unspawned_local_work_queues = self.unspawned_local_work_queues.lock().unwrap();
+        // Spawn each worker thread.
+        for (ordinal, shared) in self.workers_shared.iter().enumerate() {
+            let worker = Box::new(GCWorker::new(
+                mmtk,
+                ordinal,
+                mmtk.scheduler.clone(),
+                false,
+                sender.clone(),
+                shared.clone(),
+                unspawned_local_work_queues.pop().unwrap(),
+            ));
+            VM::VMCollection::spawn_gc_thread(tls, GCThreadContext::<VM>::Worker(worker));
+        }
+        debug_assert!(unspawned_local_work_queues.is_empty());
+    }
+
+    /// Get the number of workers in the group
+    #[inline(always)]
+    pub fn worker_count(&self) -> usize {
+        self.workers_shared.len()
+    }
+
+    /// Increase the packed-workers counter.
+    /// Called before a worker is parked.
+    ///
+    /// Return true if all the workers are parked.
+    #[inline(always)]
+    pub fn inc_parked_workers(&self) -> bool {
+        let old = self.parked_workers.fetch_add(1, Ordering::Relaxed);
```

Comment:
Similar to `pending_messages`, in theory, when a worker finds that all workers are parked, it may not see the buckets being updated.

Reply:
Yes probably either this or the update buckets code should be SeqCst, so the reordering won't happen. I'll change this.